Config instance builder

dmlloyd opened the pull request:
Update: there is now a basic (but incomplete) implementation.

This is a first draft of the builder API for constructing configuration instances. No implementation is included yet; this is just for API review.

Fixes #1001.

dmlloyd committed:
[WIP] API and implementation for config instance builder

Fixes #1001

radcortez reviewed (dismissed):
Looks good to me.

radcortez committed:
ConfigInstanceBuilder implementation:
- Builder class generation
- Simple leaf types and primitives
- Simple optionals
- @WithDefault

radcortez commented:
@dmlloyd 

I've now started to look into this and pushed some initial implementation that includes:
-  Builder class generation
- Simple leaf types and primitives
- Simple optionals
- @WithDefault

Of course, there is still a lot missing, and even the pieces I did are only for the most basic cases, so I need to start looking into all the edge cases, but it is a start.

In terms of the API, I think we could start with something smaller and drop `withString` and `withDefaultFor` variants. We could add them later if required.

On another note, `with(F getter, int value)` and `with(F getter, long value)` (as well as their `Optional` counterparts), are ambiguous when we try to call them:

```java
interface Server {
    String host();
    int port();
}

Server server = forInterface(Server.class)
        .with(Server::host, "localhost")
        .with(Server::port, 8080)
        .build();
```

So, `.with(Server::port, 8080)` is reported as ambiguous by the compiler. Casting the value type doesn't seem to help, and it seems to be related to the Function type and method reference inferred type. Would you happen to have any idea?

radcortez committed:
ConfigInstanceBuilder implementation:
- Builder class generation
- Leaf types and primitives
- Optionals
- Maps
- Collections
- @WithDefault

radcortez committed:
ConfigInstanceBuilder throw exception if a required value is not set

radcortez committed:
ConfigInstanceBuilder missing default cases:
- Optional Group
- Map with Collections
- Default methods

radcortez committed:
ConfigInstanceBuilder support @WithConverter

radcortez committed:
ConfigInstanceBuilder cleanups:
- Remove unused methods from ConfigInstanceBuilder
- Share the same code between ConfigInstanceBuilder and ConfigMappingContext
- Simplify ConfigMappingGenerator

dmlloyd commented:
Seems OK. Don't forget to JavaDoc the rest of the API.

radcortez committed:
Use modules
